[Terrain] Optimize bulk queries to the Terrain System to retrieve height, surface weights, and normals

amzn-sj opened the pull request:
The ProcessRegion/List APIs have been updated to use bulk queries. The bulk query function uses a sliding window to group positions in the same area together. This is done this way since positions that are close to each other in the list will be in same(or close) area ids. The surface weight tags are now stored in a fixed_vector instead of a regular vector to avoid reallocations during runtime.

Before:
```
Run on (64 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 524K (x32)
  L3 Unified 16777K (x8)
-----------------------------------------------------------------------------------------------------------------------
Benchmark                                                                             Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/0                                 511 ms          516 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/0                                1976 ms         1969 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/0                                7808 ms         7797 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/1                                 151 ms          148 ms            4
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/1                                 602 ms          594 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/1                                2409 ms         2422 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/2                                 150 ms          150 ms            5
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/2                                 599 ms          594 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/2                                2394 ms         2391 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/0                      458 ms          453 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/0                     1832 ms         1828 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/0                     7345 ms         7344 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/1                      129 ms          128 ms            6
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/1                      513 ms          500 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/1                     2052 ms         2047 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/2                      126 ms          125 ms            6
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/2                      506 ms          500 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/2                     2027 ms         2016 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/0                        472 ms          477 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/0                       1908 ms         1906 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/0                       7576 ms         7562 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/1                        135 ms          138 ms            5
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/1                        554 ms          562 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/1                       2195 ms         2188 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/2                        134 ms          134 ms            5
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/2                        547 ms          547 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/2                       2177 ms         2172 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/0                                1892 ms         1891 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/0                                7563 ms         7562 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/1                                 540 ms          547 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/1                                2164 ms         2172 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/2                                 535 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/2                                2144 ms         2156 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/0                     1870 ms         1859 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/0                     7473 ms         7469 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/1                      518 ms          516 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/1                     2073 ms         2078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/2                      515 ms          516 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/2                     2081 ms         2078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/0                       1852 ms         1859 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/0                       7431 ms         7422 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/1                        528 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/1                       2126 ms         2125 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/2                        521 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/2                       2100 ms         2094 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/1/2                         255 ms          255 ms            3
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/1/2                        1009 ms         1016 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/2/2                         302 ms          297 ms            2
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/2/2                        1205 ms         1203 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/4/2                         400 ms          398 ms            2
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/4/2                        1602 ms         1594 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/1/2               230 ms          229 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/1/2               921 ms          922 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/2/2               282 ms          281 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/2/2              1125 ms         1125 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/4/2               381 ms          383 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/4/2              1529 ms         1516 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/1/2                 237 ms          240 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/1/2                 967 ms          969 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/2/2                 289 ms          289 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/2/2                1183 ms         1172 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/4/2                 388 ms          391 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/4/2                1603 ms         1594 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/0                         2510 ms         2516 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/0                         9952 ms         9953 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/1                          815 ms          812 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/1                         3107 ms         3094 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/2                          771 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/2                         3080 ms         3078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/0               2422 ms         2422 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/0               9751 ms         9750 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/1                757 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/1               3020 ms         3016 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/2                746 ms          750 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/2               2984 ms         2984 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/0                 2485 ms         2484 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/0                 9832 ms         9828 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/1                  762 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/1                 3075 ms         3078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/2                  756 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/2                 3037 ms         3047 ms            1
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/1/2            0.000 ms        0.000 ms      7466667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/1/2            0.000 ms        0.000 ms      6400000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/1/2            0.000 ms        0.000 ms      7466667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/10/2           0.001 ms        0.001 ms       746667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/10/2           0.001 ms        0.001 ms       746667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/10/2           0.001 ms        0.001 ms       746667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/1000/2         0.114 ms        0.112 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/1000/2         0.114 ms        0.115 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/1000/2         0.114 ms        0.112 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/1/2         0.000 ms        0.000 ms      8960000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/1/2         0.000 ms        0.000 ms      8960000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/1/2         0.000 ms        0.000 ms      8960000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/10/2        0.001 ms        0.001 ms      1120000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/10/2        0.001 ms        0.001 ms      1120000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/10/2        0.001 ms        0.001 ms       896000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/100/2       0.006 ms        0.006 ms       112000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/100/2       0.006 ms        0.006 ms       112000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/100/2       0.006 ms        0.006 ms       112000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/1000/2      0.061 ms        0.061 ms        11200
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/1000/2      0.062 ms        0.061 ms        11200
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/1000/2      0.062 ms        0.061 ms        11200
```

After:
```
Run on (64 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 524K (x32)
  L3 Unified 16777K (x8)
-----------------------------------------------------------------------------------------------------------------------
Benchmark                                                                             Time             CPU   Iterations
-----------------------------------------------------------------------------------------------------------------------
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/0                                 492 ms          492 ms            2
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/0                                1949 ms         1953 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/0                                7721 ms         7719 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/1                                 147 ms          148 ms            4
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/1                                 589 ms          578 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/1                                2356 ms         2344 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/1024/1/2                                 146 ms          147 ms            5
TerrainSystemBenchmarkFixture/BM_GetHeight/2048/1/2                                 585 ms          594 ms            1
TerrainSystemBenchmarkFixture/BM_GetHeight/4096/1/2                                2338 ms         2344 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/0                      135 ms          134 ms            5
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/0                      566 ms          562 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/0                     2266 ms         2266 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/1                     43.5 ms         43.0 ms           16
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/1                      176 ms          176 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/1                      750 ms          750 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/1024/1/2                     42.1 ms         42.3 ms           17
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/2048/1/2                      169 ms          168 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessHeightsRegion/4096/1/2                      716 ms          719 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/0                        137 ms          138 ms            5
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/0                        586 ms          578 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/0                       2338 ms         2328 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/1                       45.4 ms         44.9 ms           16
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/1                        191 ms          193 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/1                        796 ms          797 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/1024/1/2                       42.9 ms         42.3 ms           17
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/2048/1/2                        183 ms          182 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessHeightsList/4096/1/2                        765 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/0                                1877 ms         1875 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/0                                7484 ms         7484 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/1                                 535 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/1                                2149 ms         2141 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/1024/1/2                                 537 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_GetNormal/2048/1/2                                2133 ms         2141 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/0                      573 ms          578 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/0                     2289 ms         2297 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/1                      197 ms          198 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/1                      795 ms          797 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/1024/1/2                      185 ms          184 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessNormalsRegion/2048/1/2                      753 ms          750 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/0                        580 ms          578 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/0                       2297 ms         2297 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/1                        195 ms          193 ms            3
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/1                        813 ms          812 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/1024/1/2                        188 ms          188 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessNormalsList/2048/1/2                        768 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/1/2                         254 ms          255 ms            3
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/1/2                        1011 ms         1016 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/2/2                         300 ms          297 ms            2
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/2/2                        1217 ms         1219 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/1024/4/2                         399 ms          398 ms            2
TerrainSystemBenchmarkFixture/BM_GetSurfaceWeights/2048/4/2                        1594 ms         1594 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/1/2               102 ms          102 ms            6
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/1/2               426 ms          422 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/2/2               119 ms          120 ms            6
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/2/2               509 ms          516 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/1024/4/2               157 ms          156 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsRegion/2048/4/2               662 ms          656 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/1/2                 108 ms          109 ms            6
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/1/2                 443 ms          438 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/2/2                 124 ms          122 ms            5
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/2/2                 540 ms          531 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/1024/4/2                 162 ms          160 ms            4
TerrainSystemBenchmarkFixture/BM_ProcessSurfaceWeightsList/2048/4/2                 682 ms          688 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/0                         2454 ms         2453 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/0                         9793 ms         9781 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/1                          769 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/1                         3078 ms         3078 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/1024/1/2                          762 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_GetSurfacePoints/2048/1/2                         3048 ms         3047 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/0                767 ms          766 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/0               3068 ms         3078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/1                312 ms          305 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/1               1209 ms         1203 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/1024/1/2                294 ms          289 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsRegion/2048/1/2               1156 ms         1156 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/0                  795 ms          797 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/0                 3076 ms         3078 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/1                  302 ms          297 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/1                 1245 ms         1250 ms            1
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/1024/1/2                  292 ms          297 ms            2
TerrainSystemBenchmarkFixture/BM_ProcessSurfacePointsList/2048/1/2                 1175 ms         1172 ms            1
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/1/2            0.000 ms        0.000 ms      6400000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/1/2            0.000 ms        0.000 ms      7466667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/1/2            0.000 ms        0.000 ms      6400000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/10/2           0.001 ms        0.001 ms       746667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/10/2           0.001 ms        0.001 ms       640000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/10/2           0.001 ms        0.001 ms       746667
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/100/2          0.011 ms        0.011 ms        64000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/1024/1000/2         0.113 ms        0.112 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/2048/1000/2         0.113 ms        0.112 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionRandom/4096/1000/2         0.113 ms        0.112 ms         6400
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/1/2         0.000 ms        0.000 ms     11200000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/1/2         0.000 ms        0.000 ms      8960000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/1/2         0.000 ms        0.000 ms      8960000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/10/2        0.001 ms        0.001 ms      1120000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/10/2        0.001 ms        0.001 ms      1120000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/10/2        0.001 ms        0.001 ms      1120000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/100/2       0.006 ms        0.006 ms       100000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/100/2       0.006 ms        0.006 ms       112000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/100/2       0.006 ms        0.006 ms       100000
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/1024/1000/2      0.061 ms        0.060 ms        11200
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/2048/1000/2      0.061 ms        0.060 ms        11200
TerrainSystemBenchmarkFixture/BM_GetClosestIntersectionWorstCase/4096/1000/2      0.061 ms        0.061 ms        11200
```

amzn-sj commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,163 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+        auto insertResult = areaIdToPos.insert(AZStd::pair<AZ::EntityId, TerrainHeightDataLists>(areaId, {}));
+        it = insertResult.first;
+        
+        // We maintain iterators to the positions to correspond to the input positions list.
+        // We resize these to the max expected size because we don't want any reallocations happening later.
+        // Reallocations will invalidate our iterators.
+        (*it).second.m_positions.resize(outPositions.size());
+        (*it).second.m_terrainExists.resize(outPositions.size());
```
In cases where we have multiple positions split across many area IDs, a bulk query will end up consuming a lot more memory than it needs to.

mbalfour-amzn reviewed (commented):
The overall algorithm in here is optimizing for the fewest number of bulk query calls, but at the expense of lots of data copying and allocations. On the SurfaceData side of things, data allocations have been the performance killer. It might be worth trying an alternate approach in which you just run through the list of positions once, building up a list of area id per position, then perform a bulk call for each subspan that has the same area id. 

Here's what that looks like (or might look like) in the SurfaceData code:
```
// It's possible that only some input positions will fall within this provider's AABB, but we want to query as many
// points as possible in one call, so build up subspans with as many consecutive points as possible per query.
size_t subspanBegin = 0;
size_t subspanEnd = subspanBegin;

for (size_t index = 0; index < totalQueryPositions; index++)
{
    const auto& inPosition = inPositions[index];
    SurfacePointList& surfacePointList = surfacePointLists[index];
    if (alwaysApplies || AabbContains2D(entry.m_bounds, inPosition))
        bool inBounds = hasInfiniteBounds || AabbContains2D(provider.m_bounds, inPositions[index]);
    if (inBounds)
    {
        subspanEnd++;
    }

    if ((!inBounds) || (index == (totalQueryPositions - 1)))
    {
        SurfaceDataProviderRequestBus::Event(
            entryPair.first, &SurfaceDataProviderRequestBus::Events::GetSurfacePoints, inPosition, surfacePointList);
        if ((subspanEnd - subspanBegin) > 0)
        {
            SurfaceDataProviderRequestBus::Event(
                providerHandle, &SurfaceDataProviderRequestBus::Events::GetSurfacePointsFromList,
                AZStd::span<const AZ::Vector3>(inPositions.data() + subspanBegin, inPositions.data() + subspanEnd), surfacePointList);
        }

        subspanBegin = index + 1;
        subspanEnd = subspanBegin;
    }
}
```

amzn-sj commented:
> The overall algorithm in here is optimizing for the fewest number of bulk query calls, but at the expense of lots of data copying and allocations.

So one of the reasons we need to copy data here is that in the case of BILINEAR or CLAMP samplers, we need to generate new positions from the provided input positions. Those inputs are currently passed in as const(which we can change). The `GetHeights` function writes the result into the input positions' Z component as well. Also, when calling the perPositionCallback that's provided, we need to pass in the original position and the height corresponding to that position. In the case of the bilinear sampler, we'll need to compute the final result from the intermediate results as well. While we can generate the new positions in the case of bilinear and clamp samplers and the use the span technique there, it could still result in difficult to handle cases where the intermediate results could be split across two area ids when using the bilinear sampler(? not sure about this. I'm guessing based on the existing implementation). We save memory here by not having to store the iterators but we'll still need to allocate memory for the new points being generated(and we'll have to deal with the intermediate results as well).
I see that the when retrieving surface weights, we'll end up using the input positions directly and the `GetSurfaceWeightsFromList` function call doesn't modify the input. The subspan technique should be easier to implement here.

> It might be worth trying an alternate approach in which you just run through the list of positions once, building up a list of area id per position, then perform a bulk call for each subspan that has the same area id.

In the case of the EXACT sampler, we can do this since we don't have to generate new X and Y values. But we'd still have to change the input positions list to non const if we want to do that. But that doesn't address the issue with the bilinear or clamp filters.

amzn-sj commented:
> We save memory here by not having to store the iterators but we'll still need to allocate memory for the new points being generated(and we'll have to deal with the intermediate results as well).

I can add logic to make sure that we break only on input position boundaries when using the the bilinear sampler I think. This will reduce the peak memory usage. I need to make sure to store the results for the output/callback after each bulk query. I'll try this out and check the benchmarks again.

mbalfour-amzn commented:
> So one of the reasons we need to copy data here is that in the case of BILINEAR or CLAMP samplers, we need to generate new positions from the provided input positions. 

Yeah, makes sense, there definitely needs to be _some_ copying and allocating, but I think there's more happening than necessary right now.  Here's what I'm seeing (I think):
* 1x input size:  heights vector<float>
* 1x input size:  terrainExists vector<bool>
* 1x or 4x input size:  outPositions vector<TerrainHeightDataIterators>
* Nx input size: areaIdToPos.m_positions vector<Vector3>
* Nx input size: areaIdToPos.m_terrainExists vector<bool>

Where N is the number of areas affecting the input. So even in the single area case, we have:
* 1x vector<float>
* 2x vector<bool>
* 1x vector<Vector3>
* 1x vector<TerrainHeightDataIterators>

What I'm suggesting is instead:
* vector<Vector3>
* vector<bool>
* vector<AreaId>

Where each one is 1x or 4x, depending on the sample mode.  The algorithm would look something like this:
* Create and resize the vectors
* Fill in the Vector3s with copies of the inputs, and the extra points for bilinear if needed
* Run through each position and find the best area for each one
* Call GetHeights() with subspans for the positions
* Go through the results, calculating bilinear as you go if necessary, and calling the callback

Extra areas wouldn't add any more memory, just more overhead on the "find best area", and more calls to GetHeights(). It's a time/memory tradeoff, but the memory looks to be growing by quite a bit on each area we add in the current implementation.

mbalfour-amzn commented:
(Side note - GetHeights can be changed to take in separate input and output positions if that's helpful. There's no reason it _has_ to be inOutPositions)

amzn-sj commented:
I updated the algorithm to use a sliding window approach instead. This ensures that the memory footprint remains the same regardless of which area ids the positions belong to. The down side as I mentioned in the comments is that in a scenario where the points in the list are randomly distributed, the performance will be sub optimal compared to the first implementation. But looking at the existing queries in our code base, it seems unlikely that we have a use case where the points will be randomly distributed. I updated the benchmark numbers. The performance is slightly better than the first implementation which was using more memory.

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -551,13 +713,16 @@ void TerrainSystem::ProcessHeightsFromList(
         return;
     }
 
+    AZStd::vector<bool> terrainExists(inPositions.size());
+    AZStd::vector<float> heights(inPositions.size());
```
Just as a thought, instead of heights as a vector<float>, it _might_ be faster to make this a vector<AZ::Vector3> for the positions. Instead of reserving the sizes here and passing into GetHeightsSynchronous as a span<>, you'd just pass the empty vector<>s in and let that method do the reserving.  GetHeightsSynchronous at the end would then overwrite into itself as it goes along, and then resize down to the proper size at the end.

If nothing else, it would use even less memory... :)

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,168 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    // Compute/store the final result
+    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++)
+    {
+        if (!InWorldBounds(inPositions[i].GetX(), inPositions[i].GetY()))
+        {
+            // Not in world bounds. Store defaults and move on.
+            heights[i] = minHeight;
+            terrainExists[i] = false;
+            continue;
+        }
```
This check is getting done twice for inPositions, once here and once in GenerateQueryPositions. It's probably fast enough that anything else isn't worth it, but pointing it out in case you've got any clever ideas for avoiding the double-check.

amzn-sj replied:
I fixed an edge case I missed in this [commit](https://github.com/o3de/o3de/pull/7357/commits/22e6ff72573cee957fced8edbd54fd678cfa7722). Also added a comment about it. I think we can get rid of the multiple world bounds check and instead handle it the way I do in that commit. I haven't removed the checks yet but with that commit, I think those checks are now redundant. I'll run the unit tests after getting rid of them to confirm.

amzn-sj replied:
There may be something even simpler. I can get rid of all the worlds bounds checks including the one I added for the edge case. If I initialize the z component of all the positions to min height when we're generating the positions, then the calls to GetHeights for the positions that are outside of any registered areas(this covers the outside world bounds case as well) won't get routed anywhere since the area id is invalid. This will make the code much simpler to read. With a lot of registered areas, it may be cheaper to do world bounds check. But we have to call FindClosestAreaForPosition() on all valid positions anyway. Only very few positions that are really outside bounds would have been cheaper with a bounds check. We may actually be doing much better if we remove the world bounds check and only use the FindAreaId function to filter. Because for valid positions, the world bounds check ends up being redundant.

mbalfour-amzn reviewed (approved):
This is looking really good, love that 75% speedup over the non-bulk version. :)

amzn-sj commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,186 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+                &Terrain::TerrainAreaHeightRequestBus::Events::GetHeights,
+                AZStd::span<AZ::Vector3>(outPositions.begin() + windowStart, spanLength), AZStd::span<bool>(outTerrainExists.begin() + windowStart, spanLength));
+            
+            // Sometimes, a position generated for a bilinear sample may be outside
+            // of all areas. This won't be caught by our world bounds check since we
+            // only do them on input positions and not the generated ones.
+            while (areaId == AZ::EntityId())
+            {
+                // Out of bounds. Set height to world min.
+                outPositions[i].SetZ(minHeight);
+                i++;
+                if (i < numPositions)
+                {
+                    areaId = FindBestAreaEntityAtPosition(outPositions[i].GetX(), outPositions[i].GetY(), bounds);
+                }
+                else
+                {
+                    break;
+                }
+            }
```
I found this case(in the ProcessNormalsFromList unit test) where some points generated for bilinear samples were outside all areas. In this particular unit test scenario, the world size was a lot larger than the one registered area. So the point was actually within world bounds but was outside of all areas. Is this a case that could come up in real user scenario?

amzn-sj replied:
We may be able to get rid of this. See my previous comment.

mbalfour-amzn commented on Gems/SurfaceData/Code/Include/SurfaceData/SurfaceDataTypes.h:
```diff
@@ -188,7 +182,7 @@ namespace SurfaceData
         const AzFramework::SurfaceData::SurfaceTagWeight* FindTag(AZ::Crc32 tag) const;
 
-        AZStd::fixed_vector<AzFramework::SurfaceData::SurfaceTagWeight, MaxSurfaceWeights> m_weights;
+        AZStd::fixed_vector<AzFramework::SurfaceData::SurfaceTagWeight, AzFramework::SurfaceData::Constants::MaxSurfaceWeights> m_weights;
```
Suggestion: Could just make m_weights of type AzFramework::SurfaceTagWeightList now that the two are fully aligned.

amzn-sj replied:
I'll make this change in a separate PR later

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    Sampler sampler, size_t indexStepSize) const
+{
+    const float minHeight = m_currentSettings.m_worldBounds.GetMin().GetZ();
+    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++, iteratorIndex += indexStepSize)
```
suggestion: If you use "outPositions.reserve()" instead of "outPositions.resize()" where outPositions gets defined, you could just make this loop "for (auto& inPoition : inPositions)" and use emplace_back() for pushing back each outPosition value.

amzn-sj replied:
Fixed

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    }
+}
+
+void TerrainSystem::GenerateInputPositionsFromRegion(
+    const AZ::Aabb& inRegion,
+    const AZ::Vector2& stepSize,
+    AZStd::span<AZ::Vector3> inPositions) const
```
Suggestion: Consider making this `AZStd::vector<AZ::Vector3> GenerateInputPositionsFromRegion(const AZ::Aabb& inRegion, const AZ::Vector2& stepSize)` where it creates the vector, instead of operating on it as a span<>. With the way it's written now, the inPositions vector will have to get fully filled out with default values before getting set with new ones. If you create the vector here like this:
```
AZStd::vector<AZ::Vector3> inPositions;
inPositions.reserve(numSamplesX * numSamplesY);
```
you can allocate the memory without having to construct the Vector3s, and then just use emplace_back() to push the constructed Vector3s.
Usage of the method would then look like:
```
AZStd::vector<Vector3> inPositions = GenerateInputPositionsFromRegion(inRegion, stepSize);
```

invertednormal replied:
You can solve the filling with default values problem by doing a `resize_no_construct()`, but still agree with Mike's suggestion for a more natural api. However, it'd be OK to have both. You could keep this function and make another trivial function that creates / returns a vector then uses the version that takes a span internally.

amzn-sj replied:
Updated to use Mike's suggestion

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    AZStd::vector<bool> outTerrainExists;
+
+    size_t indexStepSize = 1;
+    if (sampler == AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR)
+    {
+        // outPositions holds the iterators to results of the bulk queries.
+        // In the case of the bilinear sampler, we'll be making 4 queries per
+        // input position.
+        outPositions.resize(inPositions.size() * 4);
+        outTerrainExists.resize(inPositions.size() * 4);
+        indexStepSize = 4;
+    }
+    else
+    {
+        outPositions.resize(inPositions.size());
+        outTerrainExists.resize(inPositions.size());
+    }
```
As mentioned above, these would become "reserve()" for outPositions instead of "resize()" if you change the methods that fill them out to use emplace_back(), which would then avoid constructing a bunch of default Vector3s that will get overwritten.

amzn-sj replied:
Done

mbalfour-amzn commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -315,6 +508,38 @@ bool TerrainSystem::GetIsHoleFromFloats(float x, float y, Sampler sampler) const
+void TerrainSystem::GetNormalsSynchronous(const AZStd::span<AZ::Vector3>& inPositions, Sampler sampler, 
+    AZStd::span<AZ::Vector3> normals, AZStd::span<bool> terrainExists) const
+{
+    AZStd::vector<AZ::Vector3> directionVectors(inPositions.size() * 4);
+    const AZ::Vector2 range(m_currentSettings.m_heightQueryResolution / 2.0f, m_currentSettings.m_heightQueryResolution / 2.0f);
+    size_t indexStepSize = 4;
+    for (size_t i = 0, iteratorIndex = 0; i < inPositions.size(); i++, iteratorIndex += indexStepSize)
+    {
+        directionVectors[iteratorIndex].Set(inPositions[i].GetX(), inPositions[i].GetY() - range.GetY(), 0.0f);
+        directionVectors[iteratorIndex + 1].Set(inPositions[i].GetX() - range.GetX(), inPositions[i].GetY(), 0.0f);
+        directionVectors[iteratorIndex + 2].Set(inPositions[i].GetX() + range.GetX(), inPositions[i].GetY(), 0.0f);
+        directionVectors[iteratorIndex + 3].Set(inPositions[i].GetX(), inPositions[i].GetY() + range.GetY(), 0.0f);
+    }
```
Suggestion: Again consider reserve() and emplace_back to avoid constructing default Vector3s. It also simplifies the loop down to `for (auto& inPosition : inPositions)`

amzn-sj replied:
Fixed

invertednormal commented on Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:
```diff
@@ -177,6 +177,199 @@ bool TerrainSystem::InWorldBounds(float x, float y) const
+    AZStd::vector<AZ::Vector3> outPositions;
+    AZStd::vector<bool> outTerrainExists;
+
+    size_t indexStepSize = 1;
+    if (sampler == AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR)
+    {
+        // outPositions holds the iterators to results of the bulk queries.
+        // In the case of the bilinear sampler, we'll be making 4 queries per
+        // input position.
+        outPositions.resize(inPositions.size() * 4);
+        outTerrainExists.resize(inPositions.size() * 4);
+        indexStepSize = 4;
+    }
+    else
+    {
+        outPositions.resize(inPositions.size());
+        outTerrainExists.resize(inPositions.size());
+    }
```
Could simplify this a little bit:
```
const size_t indexStepSize = sampler == AzFramework::Terrain::TerrainDataRequests::Sampler::BILINEAR ? 4 : 1;
outPositions.reserve(inPositions.size() * indexStepSize);
outTerrainExists.reserve(inPositions.size() * indexStepSize);
```

amzn-sj replied:
Done
